casanovo/denovo/dataloaders.py

```diff
@@ -237,46 +233,81 @@ def prepare_batch(
+def prepare_psm_batch(
+    batch: List[Tuple[torch.Tensor, float, int, str]],
+    pdb: db_utils.ProteinDatabase,
+):
     """
-    Collate MS/MS spectra into a batch meant for Casanovo-DB.
+    Collate MS/MS spectra into a batch for DB search.
+
+    The MS/MS spectra will be padded so that they fit nicely as a tensor.
+    However, the padded elements are ignored during the subsequent steps.
 
     Parameters
     ----------
-    batch : List[Tuple[torch.Tensor, Tuple[float, int, float], str, Tuple[str, str]]]
+    batch : List[Tuple[torch.Tensor, float, int, str]]
         A batch of data from an AnnotatedSpectrumDataset, consisting of for each
-        spectrum (i) a tensor with the m/z and intensity peak values,
-        (ii) the precursor information [mass, charge, m/z], (iii) the
-        peptide sequence, the precursor m/z, (iv) spectrum identifiers
-        (file and scan).
+        spectrum (i) a tensor with the m/z and intensity peak values, (ii), the
+        precursor m/z, (iii) the precursor charge, (iv) the spectrum identifier.
+    pdb : db_utils.ProteinDatabase
+        The protein database to use for candidate peptide retrieval.
 
     Returns
     -------
-    spectra : torch.Tensor of shape (batch_size, n_peaks, 2)
+    all_spectra : torch.Tensor of shape (batch_size, n_peaks, 2)
         The padded mass spectra tensor with the m/z and intensity peak values
         for each spectrum.
-    precursors : torch.Tensor of shape (batch_size, 3)
+    all_precursors : torch.Tensor of shape (batch_size, 3)
         A tensor with the precursor neutral mass, precursor charge, and
         precursor m/z.
-    spectrum_peps : np.ndarray
-        Peptide sequences
-    spectrum_ids : Tuple[str, str]
-        Peak file and spectrum identifier
+    all_spectrum_ids : np.ndarray
+        The spectrum identifiers.
+    all_peptides : List[str]
+        The candidate peptides for each spectrum.
+    all_proteins : List[str]
+        The associated proteins for each candidate peptide.
     """
-    (
-        spectra,
-        precursor_mzs,
-        precursor_charges,
-        spectrum_peps,
-        spectrum_ids,
-    ) = list(zip(*batch))
+    spectra, precursor_mzs, precursor_charges, spectrum_ids = list(zip(*batch))
     spectra = torch.nn.utils.rnn.pad_sequence(spectra, batch_first=True)
+
     precursor_mzs = torch.tensor(precursor_mzs)
     precursor_charges = torch.tensor(precursor_charges)
     precursor_masses = (precursor_mzs - 1.007276) * precursor_charges
     precursors = torch.vstack(
         [precursor_masses, precursor_charges, precursor_mzs]
     ).T.float()
-    return spectra, precursors, np.asarray(spectrum_peps), spectrum_ids
+
+    all_spectra = []
+    all_precursors = []
+    all_spectrum_ids = []
+    all_peptides = []
+    all_proteins = []
+    for idx in range(len(batch)):
+        digest_data = pdb.get_candidates(
+            float(precursor_mzs[idx]),
+            float(precursor_charges[idx]),
```

Comment:
**todo:** This is `int`, not `float`. Also use the PyTorch type conversion functions instead.

Reply:
Modified